### PR TITLE
Cor 442 add yes no field type

### PR DESCRIPTION
**Pull request by neb42**
I've left the record rendering as the string "true" or "false". There should be a better visual indicator of this, but I think this is fine for the purposes of the pwa.

### CODE REVIEW

File: pkg/api/types/validation/form.go [hunk 1]
```diff
@@ -461,3 +465,9 @@ func ValidateSelectOptions(options []*types.SelectOption, path *validation.Path)
 
 	return result
 }
+
+func ValidateFieldTypeBoolean(ftBoolean *types.FieldTypeBoolean, path *validation.Path) validation.ErrorList {
```
Comment (ludydoo) [inline pkg/api/types/validation/form.go:469]: Just a thought, but perhaps that a boolean field should not be a "key" field ?

File: pkg/sqlmanager/reader.go [hunk 2]
```diff
@@ -371,6 +387,23 @@ func getIntValue(value interface{}) (types.StringOrArray, error) {
 	}
 }
 
+func getBooleanValue(value interface{}) (types.StringOrArray, error) {
+	switch t := value.(type) {
+	case bool:
+		return types.NewStringValue(strconv.FormatBool(t)), nil
```
Comment (ludydoo) [inline pkg/sqlmanager/reader.go:393]: Make sure that strconv.FormatBool does not return an uppercase "True" ;)

File: pkg/api/types/validation/form.go (hunk 1 above)
Comment (neb42) [reply pkg/api/types/validation/form.go:469]: I also think it should be required as there's no clear UI difference between false and null. In the frontend I've made it always default to false, but this isn't enforced on the backend. Do you think I should add that constraint here too?

Comment (ludydoo) [reply pkg/api/types/validation/form.go:469]: I would think so.. I can't imagine a usecase where a boolean would be part of the key.. If we get some demands for it, we could "loosen" the constraint later on?

**Review by internetti — CHANGES_REQUESTED** [review]
I cannot save a form with a checkbox in it, there's no corresponding network request
the save callback from the formerContainer gets undefined as formdefinition and the boolean field is not stored in the state with the rest of the form

File: pkg/sqlmanager/reader.go (hunk 2 above)
Comment (neb42) [reply pkg/sqlmanager/reader.go:393]: I think the test checks this

**Comment by neb42** [comment]
> I cannot save a form with a checkbox in it, there's no corresponding network request the save callback from the formerContainer gets undefined as formdefinition and the boolean field is not stored in the state with the rest of the form

Weird, I'd already fixed this but it hadn't been committed. I've pushed up a fix now.

**Review by internetti — COMMENTED** [review]
if not clicking the checkbox is automatically translated to false, why do we need to mark the field as required?

**Comment by ludydoo** [comment]
> if not clicking the checkbox is automatically translated to false, why do we need to mark the field as required?

Good point.. There is also 3 statuses for checkboxes. "Checked" "Not Checked" and "Indeterminate". The "indeterminate" one is the one with the horizontal bar in it.

https://css-tricks.com/indeterminate-checkboxes/

**Comment by neb42** [comment]
> if not clicking the checkbox is automatically translated to false, why do we need to mark the field as required?

If we didn't force it to be required this could lead to inconsistent states between the frontend and the api. If I hit the api directly I could create a checkbox with a null value, but the frontend doesn't know how to interpret it and it doesn't make much sense conceptually. By making it required we ensure that it can only ever be true or false.

**Comment by ludydoo** [comment]
Yeah it's quite ambiguous "false" or "null", but the html checkbox has support for displaying nulls differently than false. Though, very little users will know about that ! Manses sense to force required for now

**Comment by neb42** [comment]
> Good point.. There is also 3 statuses for checkboxes. "Checked" "Not Checked" and "Indeterminate". The "indeterminate" one is the one with the horizontal bar in it.
> 
> https://css-tricks.com/indeterminate-checkboxes/
> 
> Probably that we "could" (should?) allow "nulls"/"undefined" for non-required yes/no fields by allowing the "indeterminate" status? I'm thinking that in the future, when we start implementing "relevance rules" for fields, we'll need to have non-required checkboxes!

I think the use case of intermediate checkboxes isn't clear. I think they become useful if we were to implement checkbox groups, with children (think of selecting emails in gmail). So I think they fall out of the scope of this pr.

If we were to implement them I probably wouldn't use `null` as that is not what `null` means. I'd probably change the value to `0,1,2` or an enum of some sort.

**Comment by ludydoo** [comment]
Looks good to me, @internetti have you resolved the issues on your side ?

**Review by internetti — APPROVED** [review]

**Pull request merged**